{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "2f9e0027-21b8-465a-81ec-3210db5f5ad4",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/gpfs/gibbs/project/gerstein/yj329/conda_envs/py3/lib/python3.9/site-packages/pandas/core/arrays/masked.py:60: UserWarning: Pandas requires version '1.3.6' or newer of 'bottleneck' (version '1.3.5' currently installed).\n",
      "  from pandas.core import (\n"
     ]
    }
   ],
   "source": [
    "import re\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "6a79b8c3-494e-4f26-a797-2efe1d3d47c7",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.read_csv('/gpfs/gibbs/pi/gerstein/yj329/epiPgene/parent/pgene_parent_mapping_Human103.txt', sep = '\\t')\n",
    "df['PgeneEnsembl103ID'] = df['PgeneEnsembl103ID'].apply(lambda x: x.split(\".\")[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "029555ac-d97a-4cf8-8fd8-5c05e916f312",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>PseudoPipeID</th>\n",
       "      <th>Chr</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Strand</th>\n",
       "      <th>PgeneEnsembl103ID</th>\n",
       "      <th>ParentProtein</th>\n",
       "      <th>ParentGene</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>PGOHUM00000289724</td>\n",
       "      <td>chr10</td>\n",
       "      <td>80394</td>\n",
       "      <td>86265</td>\n",
       "      <td>-</td>\n",
       "      <td>ENST00000423948</td>\n",
       "      <td>ENSP00000244174</td>\n",
       "      <td>ENSG00000124334</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>PGOHUM00000289726</td>\n",
       "      <td>chr10</td>\n",
       "      <td>4871901</td>\n",
       "      <td>4891975</td>\n",
       "      <td>-</td>\n",
       "      <td>ENST00000432950</td>\n",
       "      <td>ENSP00000370240</td>\n",
       "      <td>ENSG00000187134</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>PGOHUM00000289732</td>\n",
       "      <td>chr10</td>\n",
       "      <td>6071514</td>\n",
       "      <td>6071918</td>\n",
       "      <td>-</td>\n",
       "      <td>ENST00000397237</td>\n",
       "      <td>ENSP00000380156</td>\n",
       "      <td>ENSG00000144713</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>PGOHUM00000289733</td>\n",
       "      <td>chr10</td>\n",
       "      <td>6293100</td>\n",
       "      <td>6294684</td>\n",
       "      <td>-</td>\n",
       "      <td>ENST00000440192</td>\n",
       "      <td>ENSP00000260130</td>\n",
       "      <td>ENSG00000137575</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>PGOHUM00000289735</td>\n",
       "      <td>chr10</td>\n",
       "      <td>7295106</td>\n",
       "      <td>7295290</td>\n",
       "      <td>-</td>\n",
       "      <td>ENST00000421389</td>\n",
       "      <td>ENSP00000297564</td>\n",
       "      <td>ENSG00000164919</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        PseudoPipeID    Chr    Start      End Strand PgeneEnsembl103ID  \\\n",
       "0  PGOHUM00000289724  chr10    80394    86265      -   ENST00000423948   \n",
       "1  PGOHUM00000289726  chr10  4871901  4891975      -   ENST00000432950   \n",
       "2  PGOHUM00000289732  chr10  6071514  6071918      -   ENST00000397237   \n",
       "3  PGOHUM00000289733  chr10  6293100  6294684      -   ENST00000440192   \n",
       "4  PGOHUM00000289735  chr10  7295106  7295290      -   ENST00000421389   \n",
       "\n",
       "     ParentProtein       ParentGene  \n",
       "0  ENSP00000244174  ENSG00000124334  \n",
       "1  ENSP00000370240  ENSG00000187134  \n",
       "2  ENSP00000380156  ENSG00000144713  \n",
       "3  ENSP00000260130  ENSG00000137575  \n",
       "4  ENSP00000297564  ENSG00000164919  "
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "1dbc7250-2722-4399-be64-853b628dc37f",
   "metadata": {},
   "outputs": [],
   "source": [
    "target_transcripts = set(df['PgeneEnsembl103ID'])\n",
    "target_genes = set(df['ParentGene'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e96e432f-4e79-4c02-b238-36d9fbe23394",
   "metadata": {},
   "outputs": [],
   "source": [
    "coords_db = {}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "9ea5f8d1-c6fe-4e00-8b68-434d617f7160",
   "metadata": {},
   "outputs": [],
   "source": [
    "def parse_gtf_attribute(attr_str, key):\n",
    "    start = attr_str.find(key + ' \"')\n",
    "    if start == -1: return None\n",
    "    start += len(key) + 2\n",
    "    end = attr_str.find('\"', start)\n",
    "    return attr_str[start:end]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "11e3bffd-cd3e-4db0-b45c-e486e55bd53d",
   "metadata": {},
   "outputs": [],
   "source": [
    "GTF_FILE = \"/gpfs/gibbs/pi/gerstein/yj329/epiPgene/gencode_v29/gencode.v29.annotation.gtf\"\n",
    "with open(GTF_FILE, 'r') as f:\n",
    "    for line in f:\n",
    "        if line.startswith('#'): continue\n",
    "        parts = line.strip().split('\\t')\n",
    "        if len(parts) < 9: continue\n",
    "        feature_type = parts[2] # gene or transcript\n",
    "        attributes = parts[8]\n",
    "\n",
    "        if feature_type == 'transcript':\n",
    "            tid = parse_gtf_attribute(attributes, 'transcript_id')\n",
    "            geneType =  parse_gtf_attribute(attributes, 'gene_type')\n",
    "            \n",
    "            match = re.search(r\"(processed|unprocessed)\", geneType)\n",
    "            if match:\n",
    "                geneType = match.group(1)\n",
    "                tid_clean = tid.split(\".\")[0] \n",
    "                if tid_clean in target_transcripts:\n",
    "                    coords_db[tid_clean] = {\n",
    "                        'chr': parts[0],\n",
    "                        'start': int(parts[3]),\n",
    "                        'end': int(parts[4]),\n",
    "                        'strand': parts[6],\n",
    "                        'geneType': geneType.capitalize()\n",
    "                    }\n",
    "        \n",
    "        elif feature_type == 'gene':\n",
    "            gid = parse_gtf_attribute(attributes, 'gene_id')\n",
    "            if gid:\n",
    "                gid_clean = gid.split('.')[0]\n",
    "                if gid_clean in target_genes:\n",
    "                    coords_db[gid_clean] = {\n",
    "                        'chr': parts[0],\n",
    "                        'start': int(parts[3]),\n",
    "                        'end': int(parts[4]),\n",
    "                        'strand': parts[6]}\n",
    "                    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "01ad0877-2228-458e-9008-855e461a34d2",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_upstream_1kb(fasta_obj, coord_info):\n",
    "\n",
    "    chrom = coord_info['chr']\n",
    "    strand = coord_info['strand']\n",
    "    \n",
    "    gtf_start = coord_info['start']\n",
    "    gtf_end = coord_info['end']\n",
    "    \n",
    "    seq_str = \"\"\n",
    "    \n",
    "    try:\n",
    "        if strand == '+':\n",
    "            fetch_start = gtf_start - 1 - 1000\n",
    "            fetch_end = gtf_start - 1\n",
    "            if fetch_start < 0: fetch_start = 0\n",
    "            \n",
    "            seq_str = fasta_obj.fetch(chrom, fetch_start, fetch_end)\n",
    "            \n",
    "        elif strand == '-':\n",
    "            fetch_start = gtf_end\n",
    "            fetch_end = gtf_end + 1000\n",
    "            \n",
    "            seq_str = fasta_obj.fetch(chrom, fetch_start, fetch_end)\n",
    "            seq_str = str(Seq(seq_str).reverse_complement())\n",
    "            \n",
    "    except KeyError:\n",
    "        print(f\"Error: Chromosome {chrom} not found in FASTA.\")\n",
    "        return None\n",
    "    except Exception as e:\n",
    "        print(f\"Error fetching sequence: {e}\")\n",
    "        return None\n",
    "\n",
    "    return seq_str.upper()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "f183c529-886f-4e22-ae48-5ef3b3b74731",
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio.Seq import Seq\n",
    "import pysam\n",
    "fasta = pysam.FastaFile('/gpfs/gibbs/pi/gerstein/yj329/epiPgene/gencode_v29/hg38.fa')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "27056a5c-7263-417a-9b49-2f1b995ca563",
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import Align\n",
    "\n",
    "def pairwise_alignment(seq_pgene, seq_parent, MIN_FRAG_LEN = 10):\n",
    "    aligner = Align.PairwiseAligner()\n",
    "    aligner.mode = 'local'\n",
    "    \n",
    "    aligner.match_score = 2       # Match +2\n",
    "    aligner.mismatch_score = -3   # Mismatch -3 \n",
    "    aligner.open_gap_score = -5   # Open Gap -5\n",
    "    aligner.extend_gap_score = -2 # Extend Gap -2\n",
    "        \n",
    "    alignments = aligner.align(seq_pgene, seq_parent)\n",
    "    try:\n",
    "        best_aln = alignments[0]\n",
    "    except IndexError:\n",
    "        return {\"Inclusion_Pct\": 0, \"Identity_Pct\": 0, \"Coordinates\": \"[]\"}\n",
    "\n",
    "    parent_aligned_regions = best_aln.aligned[1]\n",
    "    ## filtering on alignment\n",
    "    valid_parent_len = 0     \n",
    "    valid_regions_list = []  \n",
    "    for start, end in parent_aligned_regions:\n",
    "        fragment_len = end - start\n",
    "        if fragment_len >= MIN_FRAG_LEN:\n",
    "            valid_parent_len += fragment_len\n",
    "            valid_regions_list.append((start, end)) \n",
    "    \n",
    "    # coverage       \n",
    "    parent_total_len = len(seq_parent)    \n",
    "    inclusion_pct = (valid_parent_len / parent_total_len) * 100 if parent_total_len > 0 else 0\n",
    "    # identity\n",
    "    if valid_parent_len == 0:\n",
    "        identity_pct = 0\n",
    "    else:\n",
    "        counts = best_aln.counts()\n",
    "        matches = counts.identities \n",
    "        aln_len = best_aln.length  \n",
    "        #print(f\"Valid Len: {valid_parent_len}, Total Aln Len: {aln_len}, Matches: {matches}\")\n",
    "        identity_pct = (matches / aln_len) * 100 if aln_len > 0 else 0\n",
    "\n",
    "    return {\n",
    "        \"Inclusion_Pct\": round(inclusion_pct, 2),\n",
    "        \"Identity_Pct\": round(identity_pct, 2),\n",
    "        \"Coordinates\": str(valid_regions_list)}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "605a35db-3523-4e4b-977d-f1804d9fdd81",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "13241it [04:09, 53.16it/s]\n"
     ]
    }
   ],
   "source": [
    "from tqdm import tqdm\n",
    "\n",
    "results = []\n",
    "for idx, row in tqdm(df.iterrows()):\n",
    "    pgene_id = row['PgeneEnsembl103ID'].split('.')[0]\n",
    "    parent_id = row['ParentGene']\n",
    "    \n",
    "    pgene_coords = coords_db.get(pgene_id)\n",
    "    parent_coords = coords_db.get(parent_id)\n",
    "\n",
    "    metrics = {\"Inclusion_Pct\": 0, \"Identity_Pct\": 0, \"Coordinates\": \"[]\"}\n",
    "    \n",
    "    if pgene_coords and parent_coords:\n",
    "        seq_pgene = get_upstream_1kb(fasta, pgene_coords)\n",
    "        seq_parent = get_upstream_1kb(fasta, parent_coords)   \n",
    "        metrics = pairwise_alignment(seq_pgene, seq_parent)\n",
    "        \n",
    "        res_row = row.to_dict()\n",
    "        res_row.update(metrics)\n",
    "        res_row.update({'geneType': pgene_coords.get('geneType')})\n",
    "        results.append(res_row)    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "a6c10be8-1a9a-4617-8aab-10756b22dc0a",
   "metadata": {},
   "outputs": [],
   "source": [
    "final_df = pd.DataFrame(results)\n",
    "final_df = final_df.dropna().reset_index(drop=True)\n",
    "final_df = final_df[final_df['Inclusion_Pct']!=0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "id": "fab2a65f-3439-4af3-853f-a0057514efa7",
   "metadata": {},
   "outputs": [],
   "source": [
    "final_df.to_csv('/gpfs/gibbs/pi/gerstein/yj329/epiPgene/parent/tss1kb_comparison/tss1kb_inclusion_identity.txt', index=None, sep = '\\t')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
